### PR TITLE
Developer condition builder

### PR DESCRIPTION
**Pull request by mariuszkrzaczkowski**
## Changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

### Final checklist

- [ ] My code is written according to the guidelines found [here] (https://yetiforce.com/en/knowledge-base/documentation/developer-documentation/item/how-to-create-php-files).
- [ ] I have signed the Contributor Licence Agreement between me and YetiForce.

### CODE REVIEW

File: modules/Vtiger/views/ConditionBuilder.php [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+ */
+class Vtiger_ConditionBuilder_View extends Vtiger_IndexAjax_View
+{
+	/**
+	 * {@inheritdoc}
+	 */
+	public function checkPermission(\App\Request $request)
+	{
+		if (!Users_Privileges_Model::getCurrentUserPrivilegesModel()->hasModulePermission($request->getByType('sourceModuleName', 2))) {
+			throw new \App\Exceptions\NoPermitted('LBL_PERMISSION_DENIED', 406);
```
Comment (mariuszkrzaczkowski) [inline modules/Vtiger/views/ConditionBuilder.php]: throw new \App\Exceptions\NoPermitted('ERR_PERMISSION_DENIED', 406);

File: public_html/layouts/basic/modules/CustomView/resources/CustomView.js [hunk 2]
```diff
@@ -15,7 +15,8 @@ class CustomView {
 		let progressIndicatorElement = $.progressIndicator();
 		app.showModalWindow(null, url, () => {
 			this.contentsCotainer = $('.js-filter-modal__container');
-			this.advanceFilterInstance = Vtiger_AdvanceFilter_Js.getInstance(this.contentsCotainer.find('.filterContainer'));
+			this.advanceFilterInstance = new Vtiger_ConditionBuilder_Js(this.contentsCotainer.find('.js-condition-builder'), $('#sourceModule').val());
```
Comment (mariuszkrzaczkowski) [inline public_html/layouts/basic/modules/CustomView/resources/CustomView.js]: $('#sourceModule').val()  >>  this.contentsCotainer.find('#sourceModule').val()

File: public_html/layouts/basic/modules/Vtiger/resources/ConditionBuilder.js [hunk 3]
```diff
@@ -0,0 +1,170 @@
...
+
+	/**
+	 * Register events to add group
+	 */
+	registerAddGroup() {
+		var self = this;
+		this.container.on('click', '.js-group-add', function (e) {
+			let template = self.container.find('.js-condition-builder-group-template').clone();
+			template.removeClass('hide');
+			$(e.currentTarget).closest('.js-condition-builder-group-container').find('> .js-condition-builder-conditions-container').append(template.html());
```
Comment (mariuszkrzaczkowski) [inline public_html/layouts/basic/modules/Vtiger/resources/ConditionBuilder.js]: $(this)

File: public_html/layouts/basic/modules/Vtiger/resources/ConditionBuilder.js [hunk 4]
```diff
@@ -0,0 +1,170 @@
...
+	registerAddCondition() {
+		let self = this;
+		this.container.on('click', '.js-condition-add', function (e) {
+			let progress = $.progressIndicator({
+				position: 'html',
+				blockInfo: {
+					enabled: true
+				}
+			});
+			let container = $(e.currentTarget).closest('.js-condition-builder-group-container').find('> .js-condition-builder-conditions-container');
```
Comment (mariuszkrzaczkowski) [inline public_html/layouts/basic/modules/Vtiger/resources/ConditionBuilder.js]: $(this)

File: public_html/layouts/basic/modules/Vtiger/resources/ConditionBuilder.js [hunk 5]
```diff
@@ -0,0 +1,170 @@
...
+			$(e.currentTarget).closest('.js-condition-builder-group-container').find('> .js-condition-builder-conditions-container').append(template.html());
+		});
+	}
+
+	/**
+	 * Register events to remove group
+	 */
+	registerDeleteGroup() {
+		this.container.on('click', '.js-group-delete', function (e) {
+			$(e.currentTarget).closest('.js-condition-builder-group-container').remove();
```
Comment (mariuszkrzaczkowski) [inline public_html/layouts/basic/modules/Vtiger/resources/ConditionBuilder.js]: $(this)

File: public_html/layouts/basic/modules/Vtiger/resources/ConditionBuilder.js [hunk 6]
```diff
@@ -0,0 +1,170 @@
...
+		});
+	};
+
+	/**
+	 * Register events to remove condition
+	 */
+	registerDeleteCondition() {
+		this.container.on('click', '.js-condition-delete', function (e) {
+			$(e.currentTarget).closest('.js-condition-builder-conditions-row').remove();
```
Comment (mariuszkrzaczkowski) [inline public_html/layouts/basic/modules/Vtiger/resources/ConditionBuilder.js]: $(this)

File: app/QueryGenerator.php [hunk 7]
```diff
@@ -991,31 +1003,69 @@ public function setStateCondition($state)
...
+	 *
+	 * @throws \App\Exceptions\AppException
+	 *
+	 * @return array|bool
+	 */
+	private function getRelatedCondition(array $condition)
+	{
+		$field = $this->addRelatedJoin($condition);
+		if (!$field) {
+			Log::error('Not found source field');
```
Comment (mariuszkrzaczkowski) [inline app/QueryGenerator.php]:
```suggestion
			Log::error('Not found source field', __METHOD__);
```

File: install/install_schema/scheme.sql [hunk 8]
```diff
@@ -1,3 +1,8 @@
+/*
```
Comment (mariuszkrzaczkowski) [inline install/install_schema/scheme.sql]: to remove

File: layouts/basic/modules/CustomView/DuplicateRow.tpl [hunk 9]
```diff
@@ -0,0 +1,29 @@
+{strip}
```
Comment (mariuszkrzaczkowski) [inline layouts/basic/modules/CustomView/DuplicateRow.tpl]: no license

File: layouts/basic/modules/Vtiger/ConditionBuilder.tpl [hunk 10]
```diff
@@ -0,0 +1,8 @@
+{strip}
```
Comment (mariuszkrzaczkowski) [inline layouts/basic/modules/Vtiger/ConditionBuilder.tpl]: no license

File: layouts/basic/modules/Vtiger/ConditionBuilder/Base.tpl [hunk 11]
```diff
@@ -0,0 +1,12 @@
+{*<!-- {[The file is published on the basis of YetiForce Public License 3.0 that can be found in the following directory: licenses/LicenseEN.txt or yetiforce.com]} -->*}
+{strip}
+	<div class="tpl-ConditionBuilder-Base input-group input-group-sm">
```
Comment (mariuszkrzaczkowski) [inline layouts/basic/modules/Vtiger/ConditionBuilder/Base.tpl]:
```suggestion
	<div class="tpl-Base-ConditionBuilder-Base input-group input-group-sm">
```

File: layouts/basic/modules/Vtiger/ConditionBuilder/BaseNoValidation.tpl [hunk 12]
```diff
@@ -0,0 +1,12 @@
+{*<!-- {[The file is published on the basis of YetiForce Public License 3.0 that can be found in the following directory: licenses/LicenseEN.txt or yetiforce.com]} -->*}
+{strip}
+	<div class="tpl-ConditionBuilder-BaseNoValidation input-group input-group-sm">
```
Comment (mariuszkrzaczkowski) [inline layouts/basic/modules/Vtiger/ConditionBuilder/BaseNoValidation.tpl]:
```suggestion
	<div class="tpl-Base-ConditionBuilder-BaseNoValidation input-group input-group-sm">
```

File: layouts/basic/modules/Vtiger/ConditionBuilder/Boolean.tpl [hunk 13]
```diff
@@ -0,0 +1,16 @@
+{*<!-- {[The file is published on the basis of YetiForce Public License 3.0 that can be found in the following directory: licenses/LicenseEN.txt or yetiforce.com]} -->*}
+{strip}
+	<div class="tpl-ConditionBuilder-Boolean">
```
Comment (mariuszkrzaczkowski) [inline layouts/basic/modules/Vtiger/ConditionBuilder/Boolean.tpl]:
```suggestion
	<div class="tpl-Base-ConditionBuilder-Boolean">
```

File: layouts/basic/modules/Vtiger/ConditionBuilder/Country.tpl [hunk 14]
```diff
@@ -0,0 +1,15 @@
+{*<!-- {[The file is published on the basis of YetiForce Public License 3.0 that can be found in the following directory: licenses/LicenseEN.txt or yetiforce.com]} -->*}
+{strip}
+	{assign var=PICKLIST_VALUES value=$FIELD_MODEL->getPicklistValues()}
+	{assign var=FIELD_VALUES value=explode('##', $VALUE)}
+	<div class="tpl-ConditionBuilder-Country">
```
Comment (mariuszkrzaczkowski) [inline layouts/basic/modules/Vtiger/ConditionBuilder/Country.tpl]:
```suggestion
	<div class="tpl-Base-ConditionBuilder-Country">
```

File: layouts/basic/modules/Vtiger/ConditionBuilder/Currency.tpl [hunk 15]
```diff
@@ -0,0 +1,24 @@
+{*<!-- {[The file is published on the basis of YetiForce Public License 3.0 that can be found in the following directory: licenses/LicenseEN.txt or yetiforce.com]} -->*}
+{strip}
+	<div class="tpl-ConditionBuilder-Currency input-group input-group-sm">
```
Comment (mariuszkrzaczkowski) [inline layouts/basic/modules/Vtiger/ConditionBuilder/Currency.tpl]:
```suggestion
	<div class="tpl-Base-ConditionBuilder-Currency input-group input-group-sm">
```

File: layouts/basic/modules/Vtiger/ConditionBuilder/Date.tpl [hunk 16]
```diff
@@ -0,0 +1,18 @@
+{*<!-- {[The file is published on the basis of YetiForce Public License 3.0 that can be found in the following directory: licenses/LicenseEN.txt or yetiforce.com]} -->*}
+{strip}
+	<div class="tpl-ConditionBuilder-Date input-group input-group-sm date">
```
Comment (mariuszkrzaczkowski) [inline layouts/basic/modules/Vtiger/ConditionBuilder/Date.tpl]:
```suggestion
	<div class="tpl-Base-ConditionBuilder-Date input-group input-group-sm date">
```

File: layouts/basic/modules/Vtiger/ConditionBuilder/DateRange.tpl [hunk 17]
```diff
@@ -0,0 +1,19 @@
+{*<!-- {[The file is published on the basis of YetiForce Public License 3.0 that can be found in the following directory: licenses/LicenseEN.txt or yetiforce.com]} -->*}
+{strip}
+	<div class="tpl-ConditionBuilder-Date input-group input-group-sm date">
```
Comment (mariuszkrzaczkowski) [inline layouts/basic/modules/Vtiger/ConditionBuilder/DateRange.tpl]:
```suggestion
	<div class="tpl-Base-ConditionBuilder-Date input-group input-group-sm date">
```

File: layouts/basic/modules/Vtiger/ConditionBuilder/Modules.tpl [hunk 18]
```diff
@@ -0,0 +1,15 @@
+{*<!-- {[The file is published on the basis of YetiForce Public License 3.0 that can be found in the following directory: licenses/LicenseEN.txt or yetiforce.com]} -->*}
+{strip}
+	{assign var=PICKLIST_VALUES value=$FIELD_MODEL->getModulesListValues()}
+	{assign var=FIELD_VALUES value=explode(',', $VALUE)}
+	<div class="tpl-ConditionBuilder-Modules">
```
Comment (mariuszkrzaczkowski) [inline layouts/basic/modules/Vtiger/ConditionBuilder/Modules.tpl]:
```suggestion
	<div class="tpl-Base-ConditionBuilder-Modules">
```

File: layouts/basic/modules/Vtiger/ConditionBuilder/Owner.tpl [hunk 19]
```diff
@@ -0,0 +1,48 @@
+{*<!-- {[The file is published on the basis of YetiForce Public License 3.0 that can be found in the following directory: licenses/LicenseEN.txt or yetiforce.com]} -->*}
+{strip}
+	<div class="tpl-ConditionBuilder-Owner">
```
Comment (mariuszkrzaczkowski) [inline layouts/basic/modules/Vtiger/ConditionBuilder/Owner.tpl]:
```suggestion
	<div class="tpl-Base-ConditionBuilder-Owner">
```

File: modules/CustomView/actions/Save.php [hunk 20]
```diff
@@ -86,14 +86,14 @@ private function getCVModelFromRequest(\App\Request $request)
...
 		$customViewData['columnslist'] = $selectedColumnsList;
-		$stdFilterList = $request->get('stdfilterlist');
-		if (!empty($stdFilterList)) {
-			$customViewData['stdfilterlist'] = $stdFilterList;
-		}
 		$advFilterList = $request->get('advfilterlist');
 		if (!empty($advFilterList)) {
 			$customViewData['advfilterlist'] = $advFilterList;
 		}
+		$duplicateFields = $request->get('duplicatefields');
```
Comment (mariuszkrzaczkowski) [inline modules/CustomView/actions/Save.php]: Make validation of the received data
```suggestion
		$duplicateFields = $request->get('duplicatefields',?);
```

File: modules/Vtiger/views/ConditionBuilder.php [hunk 21]
```diff
@@ -0,0 +1,62 @@
...
+		$sourceModuleName = $request->getByType('sourceModuleName', 2);
+		$sourceModuleModel = Vtiger_Module_Model::getInstance($sourceModuleName);
+		$recordStructureModulesField = [];
+		foreach ($sourceModuleModel->getFieldsByReference() as $referenceField) {
+			foreach ($referenceField->getReferenceList() as $relatedModuleName) {
+				$recordStructureModulesField[$relatedModuleName][$referenceField->getFieldName()] = Vtiger_RecordStructure_Model::getInstanceForModule(Vtiger_Module_Model::getInstance($relatedModuleName))->getStructure();
+			}
+		}
+		if ($request->isEmpty('fieldname')) {
+			$fieldModel = reset($sourceModuleModel->getFields());
```
Comment (mariuszkrzaczkowski) [inline modules/Vtiger/views/ConditionBuilder.php]:
```suggestion
			$fieldModel = current($sourceModuleModel->getFields());
```

File: modules/Vtiger/views/ConditionBuilder.php [hunk 22]
```diff
@@ -0,0 +1,62 @@
...
+		} else {
+			$selectedOperator = $request->getByType('operator', 'Alnum');
+		}
+		$viewer = $this->getViewer($request);
+		$viewer->assign('OPERATORS', $operators);
+		$viewer->assign('SELECTED_OPERATOR', $selectedOperator);
+		$viewer->assign('SELECTED_FIELD_MODEL', $fieldModel);
+		$viewer->assign('RECORD_STRUCTURE_RELATED_MODULES', $recordStructureModulesField);
+		$viewer->assign('RECORD_STRUCTURE', Vtiger_RecordStructure_Model::getInstanceForModule($sourceModuleModel)->getStructure());
+		$viewer->assign('FIELD_INFO', $fieldInfo);
```
Comment (mariuszkrzaczkowski) [inline modules/Vtiger/views/ConditionBuilder.php]: no default value, potential notice

File: public_html/layouts/basic/modules/CustomView/resources/CustomView.js [hunk 23]
```diff
@@ -133,6 +134,38 @@ class CustomView {
...
+		const container = this.getContentsContainer();
+		App.Fields.Picklist.showSelect2ElementView(container.find('.js-duplicates-container select.js-duplicates-field'));
+		container.on('click', '.js-duplicates-remove', function(e) {
+			$(this).closest('.js-duplicates-row').remove();
+		});
+		container.find('.js-duplicate-add-field').on('click', function(){
+			let template = container.find('.js-duplicates-field-template').clone();
+			template.removeClass('d-none');
+			template.removeClass('js-duplicates-field-template');
+			App.Fields.Picklist.showSelect2ElementView(template.find('select.js-duplicates-field'));
```
Comment (mariuszkrzaczkowski) [inline public_html/layouts/basic/modules/CustomView/resources/CustomView.js]: select.js-duplicates-field ? .js-duplicates-field

File: public_html/layouts/basic/modules/Vtiger/resources/ConditionBuilder.js [hunk 24]
```diff
@@ -0,0 +1,182 @@
...
+
+
+	/**
+	 * Register events when change conditions
+	 * @param {jQuery} container
+	 */
+	registerChangeConditions(container) {
+		let self = this;
+		container.find('.js-conditions-fields, .js-conditions-operator').on('change', function (e) {
+			let progress = $.progressIndicator({
```
Comment (mariuszkrzaczkowski) [inline public_html/layouts/basic/modules/Vtiger/resources/ConditionBuilder.js]: let target = $(e.currentTarget)

File: public_html/layouts/basic/modules/Vtiger/resources/ConditionBuilder.js [hunk 25]
```diff
@@ -0,0 +1,182 @@
...
+	 * @returns {object}
+	 */
+	readCondition(container) {
+		let self = this;
+		let condition = container.find('> .js-condition-switch .js-condition-switch-value').hasClass('active') ? 'AND' : 'OR';
+		let arr = {};
+		arr['condition'] = condition;
+		let rules = [];
+		container.find('> .js-condition-builder-conditions-container >').each(function () {
+			if ($(this).hasClass('js-condition-builder-conditions-row')) {
```
Comment (mariuszkrzaczkowski) [inline public_html/layouts/basic/modules/Vtiger/resources/ConditionBuilder.js]: lat element = $(this)

File: vtlib/Vtiger/Filter.php [hunk 26]
```diff
@@ -163,95 +161,26 @@ public function addField(FieldBasic $fieldInstance, $index = 0)
...
-			'column_condition' => $condition,
-		])->execute();
-		\App\Log::trace('Adding Condition ' . self::translateComparator($comparator, true) . " on $fieldInstance->name of $this->name filter ... DONE", __METHOD__);
-
+		$cvRecordModel = \CustomView_Record_Model::getCleanInstance();
+		$cvRecordModel->set('cvid', $this->id);
+		$cvRecordModel->set('advfilterlist', $conditions);
+		$cvRecordModel->set('advfilterlistDbFormat', true);
+		$cvRecordModel->setConditionsForFilter();
+		\App\Log::trace('Adding Condition');
```
Comment (mariuszkrzaczkowski) [inline vtlib/Vtiger/Filter.php]:
```suggestion
		\App\Log::trace('Adding Condition', __METHOD__);
```

File: modules/Vtiger/uitypes/Phone.php [hunk 27]
```diff
@@ -25,6 +25,14 @@ public function getDBValue($value, $recordModel = false)
 		return \App\Purifier::decodeHtml($value);
 	}
 
+	/**
+	 * {@inheritdoc}
+	 */
+	public function getDbConditionBuilderValue($value, string $operator)
+	{
+		return $this->getDBValue($value);
```
Comment (rskrzypczak) [inline modules/Vtiger/uitypes/Phone.php]: Why skip validate()?

**Pull request merged**